Analysis/Core/include/Analysis/RecoDecay.h

```diff
@@ -388,6 +401,70 @@ class RecoDecay
     return std::sqrt(M2(args...));
   }
 
+  // Calculation of topological quantities
+
+  /// Calculates impact parameter in the bending plane of the particle w.r.t. a point
+  /// \param posP  {x, y, z} position of the point
+  /// \param posSV  {x, y, z} position of the secondary vertex
+  /// \param mom  {x, y, z} particle momentum array
+  /// \return impact parameter in {x, y}
+  template <typename T, typename U, typename V>
+  static double ImpParXY(const T& posP, const U& posSV, const array<V, 3>& mom)
+  {
+    // Ported from AliAODRecoDecay::ImpParXY
+    auto flineXY = array{posSV[0] - posP[0], posSV[1] - posP[1]};
+    auto k = dotProd(flineXY, array{mom[0], mom[1]}) / Pt2(mom);
+    auto dx = flineXY[0] - k * (double)mom[0];
+    auto dy = flineXY[1] - k * (double)mom[1];
+    auto absImpPar = sqrtSumOfSquares(dx, dy);
+    auto fline = array{posSV[0] - posP[0], posSV[1] - posP[1], posSV[2] - posP[2]};
+    auto cross = crossProd(mom, fline);
+    return (cross[2] > 0. ? absImpPar : -1. * absImpPar);
+  }
+
+  /// Calculates the difference between measured and expected track impact parameter
+  /// normalized to its uncertainty
+  /// \param decLenXY decay lenght in {x, y} plane
+  /// \param errDecLenXY error on decay lenght in {x, y} plane
+  /// \param momMother {x, y, z} or {x, y} candidate momentum array
+  /// \param impParProng prong impact parameter
+  /// \param errImpParProng error on prong impact parameter
+  /// \param momProng {x, y, z} or {x, y} prong momentum array
+  /// \return normalized difference between expected and observed impact parameter
+  template <std::size_t N, typename T, typename U, typename V, typename Z>
+  static double NormImpParMeasMinusExpProng(T decLenXY, T errDecLenXY, const array<U, N>& momMother, V impParProng, V errImpParProng, const array<Z, N>& momProng)
```

Comment:
The type templates should be more general. I don't see why `decLenXY`, `errDecLenXY` must both be `T`, same for `V` in `impParProng`, `errImpParProng` and `N` in `momMother`, `momProng`.

Reply:
I tried to avoid having a number of type templates equal to the number of variables. E. g. `decLenXY` and `errDecLenXY` are related to the same "quantity" so it makes sense that they are of the same type. However, no problem in changing it